logs/querysample/querysample.go

```diff
@@ -0,0 +1,406 @@
+			return state.PostgresQuerySample{}, err
+		}
+		sampleQueryText = util.NormalizeQuery(sampleQueryText, "unparsable", -1)
+	}
+
+	// Remove query text from EXPLAIN itself, to avoid duplication and match EXPLAIN (FORMAT JSON)
+	planDetails.QueryText = ""
+
+	explainJSON, err := json.Marshal(planDetails)
+	if err != nil {
+		return state.PostgresQuerySample{}, err
+	}
+
+	return state.PostgresQuerySample{
+		Query:         sampleQueryText,
+		RuntimeMs:     queryRuntimeMs,
+		OccurredAt:    logLine.OccurredAt,
+		Username:      logLine.Username,
+		Database:      logLine.Database,
+		LogLineUUID:   logLine.UUID,
+		HasExplain:    true,
+		ExplainSource: pganalyze_collector.QuerySample_AUTO_EXPLAIN_EXPLAIN_SOURCE,
+		ExplainFormat: pganalyze_collector.QuerySample_JSON_EXPLAIN_FORMAT,
+		// Reformat JSON so its the same as when using EXPLAIN (FORMAT JSON)
+		ExplainOutput: "[" + string(explainJSON) + "]",
+	}, nil
+}
+
+func TransformLogExplainJSON(sample state.PostgresQuerySample, normalize bool) (state.PostgresQuerySample, error) {
+	var planDetails []PlanContainer
+
+	if err := json.Unmarshal([]byte(sample.ExplainOutput), &planDetails); err != nil {
+		return sample, err
+	}
+	if len(planDetails) != 1 {
+		return sample, fmt.Errorf("Unexpected plan size: %d (expected 1)", len(planDetails))
+	}
+
+	if normalize {
+		var err error
+		planDetails[0].Plan, err = normalizePlan(planDetails[0].Plan)
+		if err != nil {
+			return sample, err
+		}
+	}
+
+	explainJSON, err := json.Marshal(planDetails)
+	if err != nil {
+		return sample, err
+	}
+
+	sample.ExplainOutput = string(explainJSON)
+
+	return sample, nil
+}
+
+var autoExplainTextPlanDetailsRegexp = regexp.MustCompile(`^Query Text: (.+)\s+([\s\S]+)`)
+
+func transformExplainTextToQuerySample(logLine state.LogLine, explainText string, queryRuntimeMs float64, normalize bool) (state.PostgresQuerySample, error) {
+	if normalize {
+		return state.PostgresQuerySample{}, fmt.Errorf("Requested 'filter_query_sample = normalize', but auto_explain format is text - not supported (discarding sample)")
+	}
+
+	explainParts := autoExplainTextPlanDetailsRegexp.FindStringSubmatch(explainText)
+	if len(explainParts) != 3 {
+		return state.PostgresQuerySample{}, fmt.Errorf("auto_explain output doesn't match expected format")
+	}
+	return state.PostgresQuerySample{
+		Query:         strings.TrimSpace(explainParts[1]),
+		RuntimeMs:     queryRuntimeMs,
+		OccurredAt:    logLine.OccurredAt,
+		Username:      logLine.Username,
+		Database:      logLine.Database,
+		LogLineUUID:   logLine.UUID,
+		HasExplain:    true,
+		ExplainSource: pganalyze_collector.QuerySample_AUTO_EXPLAIN_EXPLAIN_SOURCE,
+		ExplainFormat: pganalyze_collector.QuerySample_TEXT_EXPLAIN_FORMAT,
+		ExplainOutput: explainParts[2],
+	}, nil
+}
+
+func TransformLogMinDurationStatementToQuerySample(logLine state.LogLine, queryText string, queryRuntime string, queryProtocolStep string, parameterParts [][]string) (state.PostgresQuerySample, bool) {
```

Comment:
Maybe we should document the semantics of the boolean return value here, at least with a name?